### DISCUSSION TITLE
Custom crafting recipe for Crafting Table.

### DISCUSSION THREAD

**Discussion by Intezium** [FabricMC/fabric-api#2391] [Mod Dev Support] [answered]
I'm new to modding. I'm trying to out how to make a craft pickaxe and save dirt amount data in ExamplePickaxe class.
I've looked at the tutorial on creating custom recipes, but I don't really understand what exactly I need to do.
![ExamplePickaxe](https://user-images.githubusercontent.com/20191322/178731368-9078b96d-694b-4fcd-a728-2b36bbe99269.png)

`ExamplePickaxe.java`
```
public class ExamplePickaxe extends PickaxeItem {
    private int dirtAmount = 0;

    protected ExamplePickaxe(ToolMaterial material, int attackDamage, float attackSpeed, Settings settings) {
        super(material, attackDamage, attackSpeed, settings);
    }

    public int GetDirtAmount()
    {
        return dirtAmount;
    }
    
    @Override
    public void appendTooltip(ItemStack stack, @Nullable World world, List<Text> tooltip, TooltipContext context) {
        tooltip.add(Text.translatable("item.tutorial.example_pickaxe.tooltip", dirtAmount));
    }
}
```

**Comment by warjort** [comment, marked as answer]
You add a recipe by including a json file in your src/main/resources/data/modid/recipes
replace modid with your actual mod identifier.
See: https://fabricmc.net/wiki/tutorial:recipes for an example

This is the same as adding recipes to minecraft datapacks, except your mod is automatically a datapack.
See: https://minecraft.fandom.com/wiki/Recipe for the full documentation on vanilla recipes.

If you are using an IDE, you can also look inside the minecraft jar and find data/minecraft/recipes - it contains all the vanilla recipes, which you can use as examples.

**Reply by Intezium** [reply]
Ok, I created a recipe, but I do not know how to increase the dirtAmount counter in the ExamplePickaxe class. 
P.S. question post update.

**Reply by warjort** [reply]
Recipe's are defined in the json files with the intention that users can change them using a datapack.
What if a user decided to use stone instead of dirt, or a completely different recipe?

If you want to do something different, you will need to use something other than a crafting table that understands your mechanism.

Also you can't store data in an Item. They are singletons. There is only ever one ExamplePickaxe in the game.
You need an ItemStack to store data. ItemStacks are what you actually see in the game.

ItemStacks are 3 things:
* A reference to the item
* A count of items
* NBT data which can be anything you want

**Reply by Intezium** [reply]
> 

Maybe NBT is what I'm looking for. Is there an NBT tutorial?

**Reply by warjort** [reply]
I am the wrong person to ask about tutorials, I don't use them. :-)
I don't see one on the fabric wiki either?

The following code shows the basic usage:
````
ItemStack stack = new ItemStack(YOUR_ITEM, 1);
NBTCompound nbt = stack.getOrCreateSubNbt("yourModId"); // create your own subtag
nbt.putInt("dirtAmount", 10);
````

**Reply by warjort** [reply]
NOTE: YOUR_ITEM must be the same object you registered in Registry.ITEM

**Reply by Intezium** [reply]
Thank you very much! This is exactly what I was looking for. You are right, there is no NBT tutorial on the Fabric Wiki.
